Simplify build scripts even more and upgrade platyPS in Appveyor

bergmeister opened the pull request:
## PR Summary

This is a follow up of PR #1082 that improves it by further simplifying and tidying up the build scripts.
Especially: removing the `$framework` parameter as we can figure it out ourselves, upgrade platyPS version installed in AppVeyor for security (see [here](https://twitter.com/xvorsx/status/1053495983588265991))

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] Make sure you've added a new test if existing tests do not effectively test the code changed and/or updated documentation
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.

rjmholt commented on build.psm1:
```diff
@@ -93,14 +93,10 @@ function Start-DocumentationBuild
-    #$modInfo = new-object Microsoft.PowerShell.Commands.ModuleSpecification -ArgumentList @{ ModuleName = "platyps"; ModuleVersion = $requiredVersionOfplatyPS}
-    #if ( $null -eq (Get-Module -ListAvailable -FullyQualifiedName $modInfo))
-    if ( $null -eq (Get-Module -ListAvailable platyPS))
+    if ( $null -eq (Get-Module -ListAvailable platyPS) -or ((Get-Module -ListAvailable platyPS | Select-Object -First 1).Version -lt [version]0.12))
```
Style: Got a space on the left of the `$null` but no space on the right of the `0.12`.

`gmo -List` is being called twice here and only looks at the first module (`gmo -List` returns module in essentially random order, so that might not be the check we want).

Maybe something like:

```powershell
$platyPS = Get-Module -ListAvailable platyPS |
    Where-Object Version -ge '0.12' |
    Sort-Object Version |
    Select-Object -First 1

if (-not $platyPS)
{
    Write-Verbose -verbose "platyPS module not found or below required version of 0.12, installing the latest version."
    Install-Module -Force -Name platyPS -Scope CurrentUser
}
```

rjmholt replied:
Refactored that slightly to use a FullyQualifiedName

bergmeister replied:
Sorry, but `Get-Module -ListAvailable @{ ModuleName = 'platyPS'; ModuleVersion = '0.12.0' }` does not work on my machine using either WPS or PSCore, I will refactor it to call `gmo` only once though and sort by version

rjmholt commented on build.psm1:
```diff
@@ -191,51 +167,48 @@ function Start-ScriptAnalyzerBuild
-            $destinationDirBinaries = "$destinationDir\coreclr"
-        }
-        elseif ($PSVersion -eq '3') {
+        if ($PSVersion -eq '3') {
```
`$PSVersion` above is compared to an `[int]` rather than a `[string]` -- I know they work out the same though :)

rjmholt replied:
This `if/elseif/else` might be a good candidate for a `switch` maybe:

```powershell
switch ($PSVersion)
{
    3
    {
    }

    4
    {
    }

    5
    {
    }

    default
    {
    }
}

bergmeister replied:
Thanks, point taken. I'll change it to a switch for 3,4,5 and 6 and throw otherwise

rjmholt commented on build.psm1:
```diff
@@ -191,51 +167,48 @@ function Start-ScriptAnalyzerBuild
             Push-Location $projectRoot/Rules
-            Write-Progress "Building ScriptAnalyzer '$framework' version '${PSVersion}' configuration '${Configuration}'"
-            $buildOutput = dotnet build Rules.csproj --framework $frameworkName --configuration "${config}"
+            Write-Progress "Building ScriptAnalyzer for PSVersion '$PSVersion' using framework '$framework' and configuration '$config'"
```
I don't see a change in variables `$config` or `$Configuration` in the script -- just want to check you don't mean `$Configuration` here.

rjmholt commented on build.psm1:
```diff
@@ -191,51 +167,48 @@ function Start-ScriptAnalyzerBuild
-            Write-Progress "Building ScriptAnalyzer '$framework' version '${PSVersion}' configuration '${Configuration}'"
-            $buildOutput = dotnet build Rules.csproj --framework $frameworkName --configuration "${config}"
+            Write-Progress "Building ScriptAnalyzer for PSVersion '$PSVersion' using framework '$framework' and configuration '$config'"
+            $buildOutput = dotnet build --framework $framework --configuration "${config}"
```
Might be worth taking the `${config}` out of brackets here

bergmeister replied:
Yes, this was a James-ism ;-)

rjmholt commented:
@JamesWTruher gentle ping on the review request

bergmeister commented:
Ok, I addressed all points, merge is now only being blocked by the build failure on Linux due to the AppVeyor image update. PR #1107 has the best possible workaround for that atm, can you please review this one @rjmholt  while James is out of office?
